### PR TITLE
fix: sneak mining to work as intended

### LINKED ISSUE #1917: Mining breaks Sneak
<a href="https://github.com/Era-Lusiphur"></img></a> **Issue by [Era-Lusiphur](https://github.com/Era-Lusiphur)**
_Monday Sep 21, 2020 at 13:54:37_
_Originally opened as: project-topaz/topaz - Issue 1174_

----

**_I have:_**

- [x] searched existing issues (http://project-topaz.com/issues/) to see if the issue has already been opened
- [x] checked the commit log to see if the issue has been resolved since my server was last updated

**_Additional Information_** (Steps to reproduce/Expected behavior) **:** 

It appears as if the end of the trade process, all detectable effects are removed from the character:
https://github.com/project-topaz/topaz/blob/release/src/map/packet_system.cpp#L1362
https://github.com/project-topaz/topaz/blob/release/src/map/status_effect.h#L56

This was most likely done while thinking of trading keys to chests and coffers, which is the correct behavior for them. However, it results in Sneak breaking when trading a Pickaxe to Mine, something it shouldn't.

https://ffxiclopedia.fandom.com/wiki/Sneak_Mining

We can't simply change the flag to the existing EFFECTFLAG_INVISIBLE, because it would ruin the proper behavior for chests.

### PR DESCRIPTION
**Pull request by Flam9**
**_I affirm:_**
- [x] that I agree to Topaz Next's [Limited Contributor License Agreement](https://github.com/topaz-next/topaz/blob/release/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

^ I have not tested my code because I don't have a dev server yet. I'd like to set one up soon, but if in the meantime someone wants to try this out for me that'd be great.

**Fixes: https://github.com/topaz-next/topaz/issues/1917**

### CODE REVIEW

File: src/map/packet_system.cpp [hunk 1]
```diff
@@ -1165,7 +1165,9 @@ void SmallPacket0x032(map_session_data_t* const PSession, CCharEntity* const PCh
     if ((PTarget != nullptr) && (PTarget->id == charid))
     {
         ShowDebug(CL_CYAN "%s initiated trade request with %s\n" CL_RESET, PChar->GetName(), PTarget->GetName());
-        if (jailutils::InPrison(PChar) || jailutils::InPrison(PTarget))
+
+        // If player is invisible or if either player is in prison don't allow the trade.
+        if (PChar->StatusEffectContainer->HasStatusEffect(EFFECT_INVISIBLE)) || jailutils::InPrison(PChar) || jailutils::InPrison(PTarget))
```
Comment (Xaver-DaRed) [inline src/map/packet_system.cpp]: I think you have an extra ")"

if (
    PChar->StatusEffectContainer->HasStatusEffect(EFFECT_INVISIBLE)) <--Right here
||
    jailutils::InPrison(PChar)
||
    jailutils::InPrison(PTarget)
)

Comment (Flam9) [inline src/map/packet_system.cpp]:
```suggestion
        if (PChar->StatusEffectContainer->HasStatusEffect(EFFECT_INVISIBLE) || jailutils::InPrison(PChar) || jailutils::InPrison(PTarget))
```

**Comment by zach2good** [comment]
Content seems reasonable 👍 
This just needs testing and confirmation that it works

**Comment by Besbe** [comment]
Tested the player trade change and it works when you're invisible and start the trade, but if another player starts the trade it goes through.
Not sure if that's the intended behavior, otherwise we could add a condition for PTarget:
`PTarget->StatusEffectContainer->HasStatusEffect(EFFECT_INVISIBLE)) `

Also we might want to use `HasStatusEffectByFlag(EFFECTFLAG_INVISIBLE) ` if we want abilities like "Hide" to be also affected.

The error message is also a little inaccurate now as 316 is "_That action cannot be used in this area._"
Maybe we can use 155: "_You cannot perform that action on the specified target._"

**Comment by Flam9** [comment]
Thank you for testing it Besbe! I can definitely make that change. I wasn't sure how to find what the error codes mapped to in the codebase, where did you find 155 from for future reference?
I don't play retail, so I'll push up amendments after someone confirms the expected/correct behavior.

**Comment by Besbe** [comment]
np at all! As for the errors, I use this page: https://github.com/Windower/Lua/wiki/Message-IDs
Not sure if that info can be found somewhere else though, been working on my own server for less than 2 months

**Comment by Flam9** [comment]
> Tested the player trade change and it works when you're invisible and start the trade, but if another player starts the trade it goes through.
Not sure if that's the intended behavior, otherwise we could add a condition for PTarget:
PTarget->StatusEffectContainer->HasStatusEffect(EFFECT_INVISIBLE))

I asked a retail player and they confirmed this is the intended behavior:
image
This is by no means exhaustive research :P, but if more information presents itself we can fast-follow with a PR using the fix you suggested.

> Also we might want to use HasStatusEffectByFlag(EFFECTFLAG_INVISIBLE) if we want abilities like "Hide" to be also affected.

I'll make that change, good tip!

> The error message is also a little inaccurate now as 316 is "That action cannot be used in this area."
Maybe we can use 155: "You cannot perform that action on the specified target."
>  As for the errors, I use this page: https://github.com/Windower/Lua/wiki/Message-IDs

Thank you very much for that link! I'll make the fix now.

**Review by zach2good — APPROVED** [review]

**Pull request merged**